File: packages/types/src/team/team.ts
/**
 * Copyright (c) 2023-present Plane Software, Inc. and contributors
 * SPDX-License-Identifier: AGPL-3.0-only
 * See the LICENSE file for details.
 */

import type { TLogoProps } from "../common";

export interface ITeam {
  id: string;
  name: string;
  description: string;
  workspace: string;
  logo_props: TLogoProps;
  members_count: number;
  projects_count: number;
  created_at: string;
  updated_at: string;
}

export type TTeamWritePayload = Partial<Pick<ITeam, "name" | "description" | "logo_props">>;

export interface ITeamMember {
  id: string;
  team: string;
  member: string;
  member_email: string;
  member_display_name: string;
  member_avatar: string;
  role: number;
  created_at: string;
  updated_at: string;
}

export type TTeamMemberWritePayload = Pick<ITeamMember, "member" | "role">;

export interface ITeamProject {
  id: string;
  team: string;
  project: string;
  project_name: string;
  project_identifier: string;
  created_at: string;
  updated_at: string;
}

export type TTeamProjectWritePayload = Pick<ITeamProject, "project">;

export interface ITeamDetail extends ITeam {
  members: ITeamMember[];
  projects: ITeamProject[];
}

File: apps/api/plane/api/serializers/team.py
# Third party imports
from rest_framework import serializers

# Module imports
from plane.db.models import Team, TeamMember, TeamProject, User
from .base import BaseSerializer


class TeamMemberSerializer(BaseSerializer):
    """Serializer for team members"""
    member_email = serializers.EmailField(source='member.email', read_only=True)
    member_display_name = serializers.CharField(source='member.display_name', read_only=True)
    member_avatar = serializers.CharField(source='member.avatar', read_only=True)

    class Meta:
        model = TeamMember
        fields = [
            "id",
            "team",
            "member",
            "member_email",
            "member_display_name",
            "member_avatar",
            "role",
            "created_at",
            "updated_at",
        ]
        read_only_fields = ["id", "created_at", "updated_at"]


class TeamProjectSerializer(BaseSerializer):
    """Serializer for team projects"""
    project_name = serializers.CharField(source='project.name', read_only=True)
    project_identifier = serializers.CharField(source='project.identifier', read_only=True)

    class Meta:
        model = TeamProject
        fields = [
            "id",
            "team",
            "project",
            "project_name",
            "project_identifier",
            "created_at",
            "updated_at",
        ]
        read_only_fields = ["id", "created_at", "updated_at"]


class TeamSerializer(BaseSerializer):
    """Full team serializer with members and projects"""
    members_count = serializers.SerializerMethodField()
    projects_count = serializers.SerializerMethodField()

    class Meta:
        model = Team
        fields = [
            "id",
            "name",
            "description",
            "workspace",
            "logo_props",
            "members_count",
            "projects_count",
            "created_at",
            "updated_at",
        ]
        read_only_fields = ["id", "created_at", "updated_at"]

    def get_members_count(self, obj):
        return obj.team_members.filter(deleted_at__isnull=True).count()

    def get_projects_count(self, obj):
        return obj.team_projects.filter(deleted_at__isnull=True).count()


class TeamDetailSerializer(TeamSerializer):
    """Detailed team serializer including members and projects"""
    members = TeamMemberSerializer(source='team_members', many=True, read_only=True)
    projects = TeamProjectSerializer(source='team_projects', many=True, read_only=True)

    class Meta(TeamSerializer.Meta):
        fields = TeamSerializer.Meta.fields + ["members", "projects"]


class TeamLiteSerializer(BaseSerializer):
    """Lightweight team serializer"""

    class Meta:
        model = Team
        fields = ["id", "name", "logo_props"]
        read_only_fields = fields
